lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs

```diff
@@ -350,75 +833,690 @@ impl Client {
+            Ok(())
+        } else {
+            return Err(try_error(&format!(
+                "received invalid completion id: {}",
+                res.completion_id
+            )));
         }
-        Err(RdpError::TryError("no active device ids".to_string()))
     }
-}
-
-/// 2.2.1.1 Shared Header (RDPDR_HEADER)
-/// This header is present at the beginning of every message in sent over the rdpdr virtual channel.
-/// The purpose of this header is to describe the type of the message.
-/// https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/29d4108f-8163-4a67-8271-e48c4b9c2a7c
-#[derive(Debug)]
-struct SharedHeader {
-    component: Component,
-    packet_id: PacketId,
-}
 
-impl SharedHeader {
-    fn new(component: Component, packet_id: PacketId) -> Self {
-        Self {
-            component,
-            packet_id,
+    pub fn handle_tdp_sd_read_response<S: Read + Write>(
+        &mut self,
+        res: SharedDirectoryReadResponse,
+        mcs: &mut mcs::Client<S>,
+    ) -> RdpResult<()> {
+        debug!("received TDP: {:?}", res);
+        if let Some(tdp_resp_handler) = self
+            .pending_sd_read_resp_handlers
+            .remove(&res.completion_id)
+        {
+            let rdp_responses = tdp_resp_handler(self, res)?;
+            let chan = &CHANNEL_NAME.to_string();
+            for resp in rdp_responses {
+                mcs.write(chan, resp)?;
+            }
+            Ok(())
+        } else {
+            return Err(try_error(&format!(
+                "received invalid completion id: {}",
+                res.completion_id
+            )));
         }
     }
-    fn decode(payload: &mut Payload) -> RdpResult<Self> {
-        let component = payload.read_u16::<LittleEndian>()?;
-        let packet_id = payload.read_u16::<LittleEndian>()?;
-        Ok(Self {
-            component: Component::from_u16(component).ok_or_else(|| {
-                invalid_data_error(&format!("invalid component value {:#06x}", component))
-            })?,
-            packet_id: PacketId::from_u16(packet_id).ok_or_else(|| {
-                invalid_data_error(&format!("invalid packet_id value {:#06x}", packet_id))
-            })?,
-        })
+
+    pub fn handle_tdp_sd_write_response<S: Read + Write>(
+        &mut self,
+        res: SharedDirectoryWriteResponse,
+        mcs: &mut mcs::Client<S>,
+    ) -> RdpResult<()> {
+        debug!("received TDP: {:?}", res);
+        if let Some(tdp_resp_handler) = self
+            .pending_sd_write_resp_handlers
+            .remove(&res.completion_id)
+        {
+            let rdp_responses = tdp_resp_handler(self, res)?;
+            let chan = &CHANNEL_NAME.to_string();
+            for resp in rdp_responses {
+                mcs.write(chan, resp)?;
+            }
+            Ok(())
+        } else {
+            return Err(try_error(&format!(
+                "received invalid completion id: {}",
+                res.completion_id
+            )));
```

Comment:
```suggestion
        }
        Err(try_error(&format!(
            "received invalid completion id: {}",
            res.completion_id
        )));
```

Let's eliminate unnecessary else branches where we can.

Reply:
2e41c37aa5bd30003c916de98f4f677809471d1c